### PR TITLE
[bluetooth.govee] Govee Bluetooth Binding initial contribution

### PR DESCRIPTION
**Pull request by cpmeister**
I got some of these devices in order to try to test the new bluetooth dynamic channel generation feature I've been working on. Sadly they didn't really help, but since I already had the devices on my hands I made a quick binding for them anyway.

Signed-off-by: Connor Petty <mistercpp2000+gitsignoff@gmail.com>

### CODE REVIEW

**Comment by cpmeister** [comment]
If you consider this a test of the new binding generation script, it seems to have really botched the `bundles/pom.xml` file by adding blank lines everywhere.

File: bundles/org.openhab.binding.bluetooth.govee/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+Only two thing types are supported by this extension at the moment. More may be added later.
+
+| Thing Type ID | Description                                     |
+|---------------|-------------------------------------------------|
+| govee_h5051   | Govee Wi-Fi Temperature Humidity Monitor        |
+| govee_h5052   | Govee Temperature Humidity Monitor              |
+| govee_h5071   | Govee Temperature Humidity Monitor              |
+| govee_h5072   | Govee Temperature Humidity Monitor              |
+| govee_h5074   | Govee Mini Temperature Humidity Monitor         |
+| govee_h5075   | Govee Temperature Humidity Monitor              |
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/README.md:16]: What do you mean by "*two* thing types"?

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeDiscoveryParticipant.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * The {@link GoveeDiscoveryParticipant} handles discovery of Govee bluetooth devices
+ *
+ * @author Connor Petty - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = BluetoothDiscoveryParticipant.class)
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeDiscoveryParticipant.java:39]:
```suggestion
@Component(service = BluetoothDiscoveryParticipant.class)
```

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+			<bridge-type-ref id="roaming"/>
+			<bridge-type-ref id="bluegiga"/>
+			<bridge-type-ref id="bluez"/>
+		</supported-bridge-type-refs>
+
+		<label>Govee Thermometer</label>
+		<description>A Govee Thermometer</description>
+
+		<channels>
+			<channel id="rssi" typeId="rssi"/>
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml:18]: Is this Channel used? Same for below.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+			<channel id="humidity" typeId="govee_humidity"/>
+
+		</channels>
+
+		<config-description>
+			<parameter name="address" type="text">
+				<label>Address</label>
+				<description>Bluetooth address in XX:XX:XX:XX:XX:XX format</description>
+			</parameter>
+		</config-description>
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml:31]: You could define a config reference to get rid of the redundant config-descriptions.

**Review by fwolter — CHANGES_REQUESTED** [review]

**Comment by cpmeister** [comment]
So my initial implementation of this binding didn't actually work when I tested it, so I've had to rewrite it from scratch.
Been working and testing this for the past month and I can finally say this is ready for review.

File: bundles/org.openhab.binding.bluetooth.govee/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+|------------------------|-------------------------|----------|-------------|---------------------------------------------------------------------------|
+| all                    | address                 | yes      |             | The Bluetooth address of the device (in format "XX:XX:XX:XX:XX:XX")       |
+| all                    | refreshInterval         |          | 300         | How often, in seconds, the sensor data of the device should be refreshed  |
+| goveeHygrometer*       | temperatureCalibration  | no       |             | Offset to apply to temperature sensor readings                            |
+| goveeHygrometer*       | humidityCalibration     | no       |             | Offset to apply to humidity sensor readings                               |        
+| goveeHygrometerMonitor | temperatureWarningAlarm |          | false       | Enables warning alarms to be broadcast when temperature is out of range   |
+| goveeHygrometerMonitor | temperatureWarningMin   |          | 0           | The lower safe temperature threshold (only applies if alarm is enabled)   |
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/README.md:29]: Unit

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+        }, 1, 2, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        try {
+            throw new Exception();
+        } catch (Exception ex) {
+            logger.debug("Disposing Govee Hygrometer {} model: {}", address, model, ex);
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java:145]: What's the purpose of this?

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetTemHumCommand.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @author Connor Petty - Initial Contribution
+ *
+ */
+@NonNullByDefault
+public class GetTemHumCommand extends GetCommand {
+
+    private CompletableFuture<@Nullable TemHumDTO> resultHandler;
+    // private BiConsumer<@Nullable QuantityType<Temperature>, @Nullable QuantityType<Dimensionless>> resultHandler;
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetTemHumCommand.java:33]: Can this be removed?

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+		<state readOnly="true" pattern="%.1f %unit%"/>
+	</channel-type>
+
+	<channel-type id="govee-temperature-alarm">
+		<item-type>Switch</item-type>
+		<label>Temperature Warning Alarm</label>
+		<description>If temperature warnings are enabled, then this alarm indicates
+			whether the current temperature is out of
+			range
+		</description>
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml:181]: Strange line break.

File: bundles/org.openhab.binding.bluetooth.govee/src/test/java/org/openhab/binding/bluetooth/govee/internal/readme/ThingTypeTableGenerator.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+import org.openhab.binding.bluetooth.govee.internal.GoveeModel;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+/**
+ * @author Connor Petty - Initial contribution
+ *
+ */
+public class ThingTypeTableGenerator {
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/test/java/org/openhab/binding/bluetooth/govee/internal/readme/ThingTypeTableGenerator.java:40]: I like this. Love to see this at some more prominent location.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+ *
+ */
+@NonNullByDefault
+public class GoveeHygrometerHandler extends ConnectedBluetoothHandler {
+
+    private static final UUID SERVICE_UUID = UUID.fromString("494e5445-4c4c-495f-524f-434b535f4857");
+    private static final UUID PROTOCOL_CHAR_UUID = UUID.fromString("494e5445-4c4c-495f-524f-434b535f2011");
+    private static final UUID KEEP_ALIVE_CHAR_UUID = UUID.fromString("494e5445-4c4c-495f-524f-434b535f2012");
+
+    private static byte[] SCAN_HEADER = { (byte) 0xFF, (byte) 0x88, (byte) 0xEC };
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java:75]:
```suggestion
    private static final byte[] SCAN_HEADER = { (byte) 0xFF, (byte) 0x88, (byte) 0xEC };
```

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetCommand.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package org.openhab.binding.bluetooth.govee.internal.command.hygrometer;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * @author Connor Petty - Initial Contribution
+ *
+ */
+public abstract class GetCommand extends GoveeCommand {
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetCommand.java:21]:
```suggestion
@NonNullByDefault
public abstract class GetCommand extends GoveeCommand {
```

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/SetCommand.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.bluetooth.govee.internal.command.hygrometer;
+
+/**
+ * @author Connor Petty - Initial Contribution
+ *
+ */
+public abstract class SetCommand extends GoveeCommand {
```
Comment (fwolter) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/SetCommand.java:19]:
```suggestion
@NonNullByDefault
public abstract class SetCommand extends GoveeCommand {
```

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.bluetooth.govee/README.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+## Thing Configuration
+
+Govee things have the following configuration parameters:
+
+| Thing                  | Parameter               | Required | Default     | Description                                                               |
+|------------------------|-------------------------|----------|-------------|---------------------------------------------------------------------------|
+| all                    | address                 | yes      |             | The Bluetooth address of the device (in format "XX:XX:XX:XX:XX:XX")       |
+| all                    | refreshInterval         |          | 300         | How often, in seconds, the sensor data of the device should be refreshed  |
+| goveeHygrometer*       | temperatureCalibration  | no       |             | Offset to apply to temperature sensor readings                            |
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/README.md:26]: The asterisk means that the config is available to both `goveeHydrometer` and the `goveeHydrometerMonitor` thing types. I'll make it more explicit in the readme.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java (hunk 6 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java:145]: wow, I totally forgot to remove that. It was strictly for debugging.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml (hunk 8 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml:181]: blame spotless

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/ConnectedBluetoothHandler.java [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+    private final Condition serviceDiscoveryCondition = deviceLock.newCondition();
+    private final Condition charCompleteCondition = deviceLock.newCondition();
+
+    private @Nullable Future<?> reconnectJob;
+    private @Nullable Future<?> pendingDisconnect;
+    private @Nullable BluetoothCharacteristic ongoingCharacteristic;
+    private @Nullable BluetoothCompletionStatus completeStatus;
+
+    private boolean connectOnDemand;
+    private int idleDisconnectDelay = 1000;
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/ConnectedBluetoothHandler.java]: done

File: bundles/org.openhab.binding.bluetooth.govee/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/README.md]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/ConnectedBluetoothHandler.java [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+        // the connectionTaskExecutor first
+        ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1,
+                new NamedThreadFactory("bluetooth-connection-" + thing.getThingTypeUID(), true));
+        executor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        executor.setRemoveOnCancelPolicy(true);
+        connectionTaskExecutor = executor;
+
+        super.initialize();
+
+        // Object connectOnDemandRaw = getConfig().get(BluetoothBindingConstants.CONFIGURATION_CONNECT_ON_DEMAND);
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/ConnectedBluetoothHandler.java]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/ConnectedBluetoothHandler.java [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+            return null;
+        });
+    }
+
+    public CompletableFuture<@Nullable Void> writeCharacteristic(UUID serviceUUID, UUID characteristicUUID, byte[] data,
+            boolean enableNotification) {
+        return executeWithConnection(serviceUUID, characteristicUUID, characteristic -> {
+            if (enableNotification) {
+                if (!device.enableNotifications(characteristic)) {
+                    // logger.debug("Failed to enable notifications for characteristic {}", characteristic.getUuid());
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/ConnectedBluetoothHandler.java]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java (hunk 6 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml (hunk 8 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml]: tried my best, but there is definitely some kinds of spotless formatting bug at play.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java (hunk 10 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/GoveeHygrometerHandler.java]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetCommand.java (hunk 11 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetCommand.java:23]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/SetCommand.java (hunk 12 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/SetCommand.java:22]: done

File: bundles/org.openhab.binding.bluetooth.govee/README.md (hunk 5 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/README.md]: added

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetTemHumCommand.java (hunk 7 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetTemHumCommand.java]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/feature/feature.xml [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+	terms of the Eclipse Public License 2.0 which is available at
+	http://www.eclipse.org/legal/epl-2.0
+
+	SPDX-License-Identifier: EPL-2.0
+
+-->
+<features name="org.openhab.binding.bluetooth.govee-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-bluetooth-govee" description="Bluetooth Binding Govee" version="${project.version}">
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/feature/feature.xml:19]: That is kinda the case for all of the bluetooth bindings...
"extension" refers to the fact that the binding adds functionality to the base bluetooth bindings.
"adapter"  is used the pom.xml as part of the project name. But strictly speaking I don't really think this is appropriate and the word should probably be changed to `extension` instead. But this standard was in place for all of the other bluetooth bindings so if I'm going to change it it should be a separate PR. 

I do agree with you that these naming conventions are a bit confusing at this point and I would love to change them.

**Review by fwolter — APPROVED** [review]
LGTM

**Comment by cpmeister** [comment]
@openhab/add-ons-maintainers This needs a second reviewer if someone can take a look. Thanks.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+			<parameter name="refreshInterval" type="integer" unit="s">
+				<label>Refresh Interval</label>
+				<description>The frequency at which battery, temperature, and humidity data will refresh</description>
+				<default>300</default>
+				<advanced>true</advanced>
+			</parameter>
+
+			<parameter name="temperatureCalibration" type="decimal" min="-1.6" max="1.6" groupName="calibration"
+				unit="Cel">
+				<required>false</required>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml:48]: element `required` is deprecated. Use attribute. But in this case as it's the same as the default value you can just remove it. Same for other parameter.

File: bundles/org.openhab.binding.bluetooth.govee/src/test/java/org/openhab/binding/bluetooth/govee/internal/readme/ThingTypeTableGenerator.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+        System.out.println(writer.toString());
+    }
+
+    private static String writeRow(int[] maxColumns, String[] row, char paddingChar) {
+        String prefix = "|" + paddingChar;
+        String infix = paddingChar + "|" + paddingChar;
+        String suffix = paddingChar + "|";
+
+        return Stream.of(0, 1, 2).map(i -> StringUtils.rightPad(row[i], maxColumns[i], paddingChar))
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.bluetooth.govee/src/test/java/org/openhab/binding/bluetooth/govee/internal/readme/ThingTypeTableGenerator.java:94]: Maybe not use apache `StringUtils` as we want to get ride of it.

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Just some small stuff.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/feature/feature.xml [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+	See the NOTICE file(s) distributed with this work for additional
+	information.
+
+	This program and the accompanying materials are made available under the
+	terms of the Eclipse Public License 2.0 which is available at
+	http://www.eclipse.org/legal/epl-2.0
+
+	SPDX-License-Identifier: EPL-2.0
+
+-->
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/feature/feature.xml]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetOrSetHumCaliCommand.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+ * @author Connor Petty - Initial Contribution
+ *
+ */
+@NonNullByDefault
+public class GetOrSetHumCaliCommand extends GoveeCommand {
+
+    private final CompletableFuture<@Nullable QuantityType<Dimensionless>> resultHandler;
+    private final @Nullable QuantityType<Dimensionless> value;
+
+    // private @Nullable Integer value;
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/java/org/openhab/binding/bluetooth/govee/internal/command/hygrometer/GetOrSetHumCaliCommand.java]: Thanks for finding that. Removed.

File: bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml (hunk 18 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/main/resources/OH-INF/thing/thing-types.xml]: done

File: bundles/org.openhab.binding.bluetooth.govee/src/test/java/org/openhab/binding/bluetooth/govee/internal/readme/ThingTypeTableGenerator.java (hunk 19 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.govee/src/test/java/org/openhab/binding/bluetooth/govee/internal/readme/ThingTypeTableGenerator.java]: done

**Review by Hilbrand — APPROVED** [review]
LGTM

**Pull request merged**